src/vs/platform/terminal/common/capabilities/bufferMarkCapability.ts

```diff
@@ -0,0 +1,53 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Emitter } from 'vs/base/common/event';
+import { IBufferMarkDetectionCapability, TerminalCapability } from 'vs/platform/terminal/common/capabilities/capabilities';
+import { IMarkProperties } from 'vs/platform/terminal/common/terminalProcess';
+// Importing types is safe in any layer
+// eslint-disable-next-line code-import-patterns
+import type { IMarker, Terminal } from 'xterm-headless';
+
+/**
+ * Stores marks to be added to the buffer
+ * anonymous marks are those that have no ID
+ */
+export class BufferMarkCapability implements IBufferMarkDetectionCapability {
+
+	readonly type = TerminalCapability.BufferMarkDetection;
+
+	private _idToMarkerMap: Map<string, IMarker> = new Map();
+	private _anonymousMarkers: IMarker[] = [];
+
+	private readonly _onMarkAdded = new Emitter<IMarkProperties>();
+	readonly onMarkAdded = this._onMarkAdded.event;
+
+	constructor(
+		private readonly _terminal: Terminal
+	) {
+	}
+
+	markers(): IMarker[] { return Array.from(this._idToMarkerMap.values()).concat(this._anonymousMarkers); }
```

Comment:
An iterator would be better here to avoid creating extra arrays:

```ts
for (const m of this._idToMarkerMap) {
  yield m;
}
for (const m of this._anonymousMarkers) {
  yield m;
}
```